zebROS_ws/src/tf_object_detection/src/object_detection.cpp

```diff
@@ -26,30 +29,288 @@ void camera_info_callback(const sensor_msgs::CameraInfoConstPtr &info)
+	[ INFO] [1635948302.568242461]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/goal_with_power_cell_in_front.jpg
+	[ INFO] [1635948302.598558216]: Calculated depth is 0.001 ❌
+	[ INFO] [1635948312.950812216]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/test_weirdness.jpg
+	[ INFO] [1635948312.980837687]: Calculated depth is 73.001 ✅
+	*/
+}
+
+// Get the most useful depth value in the cv::Mat depth contained within
+// the supplied bounding rectangle, using k-means
+double kMeansDepthMat(const cv::Mat& depth, const cv::Rect& bound_rect, bool debug = false, int maximumK = 3, float tolerance = 1e-3)
 {
-	double sum = 0.0;
-	unsigned count = 0;
-	for (int j = bound_rect.tl().y+1; j < bound_rect.br().y; j++) //for each row
-	{
-		const float *ptr_depth = depth.ptr<float>(j);
+	// setup randomizing (for initialization of k-means)
+	std::random_device seeder;
+	std::mt19937 engine(seeder());
+
+	// Start of checking if k is too high
 
-		for (int i = bound_rect.tl().x+1; i < bound_rect.br().x; i++) //for each pixel in row
+	// Convert cropped depth matrix to vector
+	std::vector<float> vec;
+	cv::Mat mat = depth(bound_rect);
+	for (int i = 0; i < mat.rows; i++) {
+    vec.insert(vec.end(), mat.ptr<float>(i), mat.ptr<float>(i)+mat.cols*mat.channels());
+  }
+
+	// Find unique (defined as >1cm away) values in the vector
+	std::sort(vec.begin(), vec.end());
+	auto last = std::unique(vec.begin(), vec.end(), [](float first, float second){ return fabs(second - first)<0.01; });
+	vec.erase(last, vec.end());
+
+	// Set k to maximumK if there are more than maximumK unique values, otherwise set it to the number of unique values
+	int k = std::min((int)(vec.size()), maximumK);
+
+	// End of checking if k is too high
+
+	// initialize arrays (and a vector) for k-means
+	double centroids[k];
+	double prevCentroids[k];
+	std::vector<float> clusters[k];
+
+	// Pick centroids
+	std::sample(vec.begin(), vec.end(), centroids, k, engine);
+
+	// Print centroids
+	if (debug) {
+		ROS_INFO_STREAM("k = " << k);
+		ROS_INFO_STREAM("Centroids: ");
+		for (int i = 0; i < k; i++) {
+			ROS_INFO_STREAM(centroids[i]);
+		}
+	}
+
+	while (true) { // once the algorithm converges this returns
+		for (int j = bound_rect.tl().y+1; j < bound_rect.br().y; j++) // for each row
 		{
-			if (!(isnan(ptr_depth[i]) || isinf(ptr_depth[i]) || (ptr_depth[i] <= 0)))
+			const float *ptr_depth = depth.ptr<float>(j);
+
+			for (int i = bound_rect.tl().x+1; i < bound_rect.br().x; i++) // for each pixel in row
 			{
-				sum += ptr_depth[i];
-				count += 1;
+				if (!(isnan(ptr_depth[i]) || isinf(ptr_depth[i]) || (ptr_depth[i] <= 0)))
+				{
+					// Calculate which centroid/mean the current pixel is closest to
+					float diffs[k];
+					for (int c = 0; c < k; c++) {
+						diffs[c] = abs(centroids[c] - ptr_depth[i]);
+					}
+					int closestCentroid = std::distance(diffs, std::min_element(diffs, diffs+k));
```

Comment:
Since you're already iterating over each of the centroids, you could just track the min value and index instead of storing the diffs and then iterating over them again.  That is something like (pseudo-code)
closest_centroid_index = whatever
closest_distance = huge number
for (c in k)
{
   if (fabs(centroid[c] - depth[i]) < closest_distance)
   {
      closest_centroid_index = c
      closest_distance = fabs(c-d)
   }
}

Reply:
Ok, thanks! I changed it to this:
```cpp
// Calculate which centroid/mean the current pixel is closest to
size_t closestCentroidIndex = -1;
float smallestDiff = std::numeric_limits<float>::max();
for (size_t c = 0; c < k; c++) {
  float diff = fabs(centroids[c] - ptr_depth[i]);
  if (diff < smallestDiff) {
	  closestCentroidIndex = c;
	  smallestDiff = diff;
  }
}
// Append the pixel's value to the cluster corresponding to that centroid
clusters[closestCentroidIndex].push_back(ptr_depth[i]);